[Improvement] CompletionHandler support snippets

camilledejoye opened the pull request:
This is an improvement aiming to handle snippets during completion as explained in the [specification](https://microsoft.github.io/language-server-protocol/specification#textDocument_completion).

Works like a charm on my side.

#### Todo list:

* [x] Handle CompletionClientCapabilities.snippetSupport
* [ ] Update `textEdit` to use the value of `insertText`
* [ ] ~~Handle `InputText` for all completion kinds~~
* [ ] Differentiate completion kind `class` from `constructor`

##### Handle CompletionClientCapabilities.snippetSupport

Don't generate snippets if CompletionClientCapabilities.snippetSupport is falsy ?
Not sure if we MUST prevent snippet if the client does not handle them or not.
Given the client will use `label` only if `insertText` if falsy, I guess it's mandatory.
Except if the client check `insertTextFormat` first and fallback to `label` if it's a snippet.
But it will be cleaner to not generate snippets if the client can't handle them.

##### Update `textEdit` to use the value of `insertText`

> The `insertText` is subject to interpretation by the client side.
Some tools might not take the string literally. For example
VS Code when code complete is requested in this example `con<cursor position>`
and a completion item with an `insertText` of `console` is provided it
will only insert `sole`. Therefore it is recommended to use `textEdit` instead
since it avoids additional client side interpretation.

Currently `textEdit` is built from `Suggestion.name` it should use the same value as `insertText` given it's also suppose to handle snippet:

> The format of the insert text. The format applies to both the `insertText` property and the `newText` property of a provided `textEdit`.

##### ~~Handle `InputText` for all completion kinds~~
~~Basically it's moving the logic from `SuggestionNameFormatter` into the SuggestionInsertTextFactory`.~~
~~This way we would separate the text that will be inserted by the client from what we show in the completion list.~~
~~This might event gave us the opportunity to use glyphs with the name if vim is capable of handling them.~~

##### Differentiate completion kind `class` from `constructor`

It's probably more related to the completion package than just the `CompletionItem`.
When debugging with CoC I had this result for completion:
```json
[
    {
        "label": "Suggestion",
        "kind": 7,
        "detail": "Phpactor\\Completion\\Core\\Suggestion",
        "documentation": "",
        "sortText": null,
        "filterText": null,
        "insertText": null,
        "textEdit": null,
        "additionalTextEdits": null,
        "command": null,
        "data": null,
        "insertTextFormat": 2
    },
    {
        "label": "Suggestion",
        "kind": 7,
        "detail": "Phpactor\\Completion\\Core\\Suggestion(string $name, string $type = NULL, string $shortDescription = NULL, string $classImport = NULL, string $label = NULL, string $documentation = NULL, Range $range = NULL),",
        "documentation": "",
        "sortText": null,
        "filterText": null,
        "insertText": null,
        "textEdit": null,
        "additionalTextEdits": null,
        "command": null,
        "data": null,
        "insertTextFormat": 2
    }

]
```

I think it would be great if we can manage in the second case to declare it with kind `4` which stands for `constructor`.
This way the snippet will be generated in the same way as for a function and I hope we could enjoy it when instantiating objects with `new` :)

But I think I will keep this one as an improvement for another PR.

camilledejoye committed:
Add Protocol\InsertText to deal with snippets on completion

https://microsoft.github.io/language-server-protocol/specification#textDocument_completion

It's possible to provide the key `insertText` to let the client know
what to insert in place of the pattern typed by the user on completion.
For example:
```
new SuggesITF<>
```

Would be resolve to `SuggestionInsertTextFactory` but the client need to
replace `SuggesITF` by it.
`insertText` is used for that purpose, if `insertText` is not provided
`label` is used instead.

Another key allows to specify the type of the text to insert:
`insertTextFormat`
Which can have two values:
1. Plain text
2. Snippet

To deal with that I added a new prorocol value object `InsertText` which
provide both the text to insert and the type.
I also created a factory that can create an `InsertText` from a
`Suggestion`: `SuggestionInsertTextFactory`

The next step is to inject this factory in the `CompletionHandler` and
generate the appropriate `CompletionItem`
I would like to rewwork the `CompletionItem` in order to provide it only
the `InsertText` object instead of the value and type separatly.
Especially because they are far from each other which is impractical.
This will also reduce the knowledge we have about constructing an
`CompletionItem` by letting the object itself defining its values.

camilledejoye committed:
Support snippets in CompletionHandler

Taluu commented on lib/LanguageServerCompletion/Protocol/InsertText.php:
```diff
@@ -0,0 +1,65 @@
+     */
+    private $type;
+
+    public function __construct(?string $value, ?int $type = null)
```
this should be private IMO if you're using named constructors (so no `null` values for the type)

camilledejoye replied:
For me they are not exclusive, I also do `new InsertText(null)`.
I could add `InsertText::createEmpty()` or `InsertText::none()` but I'm not sure about the why ?

But I would love to know why so don't hesitate to expand on that ;)

Anyway I won't do it quit soon, after talking with @dantleech we decided to move a big part of this to the completion package and deal with it in `Suggestion` instead.
So I will rewrite this PR once this will be done.

camilledejoye committed:
Update regex detecting a snippet

https://microsoft.github.io/language-server-protocol/specification#snippet-syntax

The placeholders can be of the form `$1` when I only detected `${1}`.
I also added a lookbehind pattern to make sure that the `$` sign is not
escaped so that `test(\$1)` is not interpreted as a snippet.

Regex tests can be found here:
https://regex101.com/r/xkL55J/2

camilledejoye committed:
Use the new Suggestion::snippet() instead of creating it ourself

Remove everything related to snippet creation that is not needed anymore
and update the container.

camilledejoye committed:
Add a gitignore

camilledejoye committed:
Fix issue with $ for variable

Because I used $suggestion->name() as insertText the problem of the $
sign came back.
So I just needed to use the formated name instead to solve the problem.

camilledejoye commented on lib/LanguageServerCompletion/Handler/CompletionHandler.php:
```diff
@@ -90,16 +91,23 @@ public function completion(TextDocumentItem $textDocument, Position $position, C
+
                 $completionList->items[] = new CompletionItem(
-                    $this->suggestionNameFormatter->format($suggestion),
+                    $name,
```
Should we use `$suggestion->label()` instead ?
It would be more appropriate since it's what `CompletionItem` expect.
And it won't break anything since we will always have either `insertText` or `textEdit` provided, so `label` will never be used by the client to update the buffer.

dantleech replied:
the `label` in Phpactor is probably different from the `name` in some cases (not sure, but can only imagine that I added it for this reason). So I think safer to leave as name

camilledejoye replied:
I proposed that because in this PR I **always** provide an `insertText`.
For those who (whom ?) don't have the spec in mind this is a summary of what a `CompletionItem` contains:
* `label`: The label of this completion item - mandatory
* `insertText`: A string that should be inserted into a document when selecting this completion
* `insertTextFormat`: The format of the insert text - defaults to `InsertTextFormat.PlainText`
* `textEdit`: An edit which is applied to a document when selecting this completion. When an edit is provided the value of `insertText` is ignored.

So the `label` is mandatory and will only be used as the text to insert in the buffer in last resort.
For the client there is no difference between using `label` or `insertText`, the exact same logic applies.
Given this two things I decided to always provide the `textEdit` using this logic:
* if `Suggestion::snippet()` is not empty use it and set `insertTextFormat` to `InsertTextFormat.Snippet`
* otherwise use `$name` (formatted to ignore the `$` if needed) and set `insertTextFormat` to `InsertTextFormat.PlainText`

Which guarantee us that the client will never use `CompletionItem.label` as text to insert in the buffer.
So we can put whatever we want in it to help identify the item in the completion list: markers like `$`, `()`, `::` or even glyph/icon.

So I think it's pretty safe to do it, because we won't change the behavior, only what is shown in the popup.

dantleech replied:
> For those who (whom ?

`who` :) `to whom it may concern`, `for whom the bell tolls` (maybe imagine it as `whom => for the person that` .. maybe.

In anycase it's very rarely used in daily language.

dantleech replied:
I think it's still better to leave it as `name`, as it's the _client's_ choice if it uses (or can use) `insertText`. `name` is guaranteed to be correct in the "last case".

camilledejoye replied:
Since the spec says that the client *should* user `insertText` and not *must* use it, I'm fine letting it like that.
But Ii still feel like we private our self from some nice features just to respond to a broken client.
Handling `insertText` is done the same way has handling `label`.

dantleech commented on tests/LanguageServerCompletion/Unit/Handler/CompletionHandlerTest.php:
```diff
@@ -75,8 +77,8 @@ public function testHandleSuggestions()
-            new CompletionItem('hello'),
-            new CompletionItem('goodbye'),
+            new CompletionItem('hello', null, null, null, null, null, 'hello', null, null, null, null, 1),
+            new CompletionItem('goodbye', null, null, null, null, null, 'goodbye', null, null, null, null, 1),
```
Tempted to suggest using something like:

```
Invoke::new(CompletionItem::class, [ 'label' => 'goodbye', 'textEdit' => 'goodbye'])
```
https://packagist.org/packages/dantleech/invoke
(actually already used by the language server)

camilledejoye replied:
Ugly but don't mind for test purposes :)

camilledejoye committed:
Register client capabilities as param in the container

camilledejoye committed:
Use Invoke::new to create CompletionItem

camilledejoye committed:
Handle clientCapabilities.textDocument.completion.supportSnippet

dantleech commented on .gitignore:
```diff
@@ -0,0 +1,7 @@
+/tests/LanguageServerIndexer/Workspace
+/tests/Assets/workspace
+/composer.lock
+/.php_cs.cache
```
:+1: for some reason I had this locally but not in git.

dantleech replied:
actually it's a conflict now as I added it ....

dantleech commented on lib/LanguageServer/Handler/PhpactorHandlerLoader.php:
```diff
@@ -56,7 +58,9 @@ protected function createContainer(string $rootUri, array $config): Container
             $extensionClasses,
-            array_merge($parameters, $config)
+            array_merge($parameters, $params->initializationOptions, [
+                LanguageServerExtension::PARAM_CLIENT_CAPABILITIES => $params->capabilities
```
:+1: (i think) technically this is wrong as it presents an "API" to the user (they are here able to add the capabilities in the config, which will get overridden here), but we do it in other places, and could rather be improved in the "container/extension" package.

camilledejoye replied:
> they are here able to add the capabilities in the config

You mean `initializationOptions` ?
If so then I just have to inverse the order no ?

And I personally understoof `initializationOptions` as being options related to the server not the client, but the spec is light about that.

camilledejoye replied:
I just realized, should not I use `array_merge_recursive` instead ?
Since the capabilities are multiple dimensional array.
Or we need to "normalize" it in order to recreate keys like `textDocument.completion`.
Which I think I will do anyway in a new service `ClientCapabilitiesProvider`

The main reason for that service is that I don't like to handle array for starter, there is no way to define a contract about what is in it.
Second we are forced to either know it's structure in the container or provide it as a hole to a service requiring one of the client capabilities.
I feel like the container should not be the one knowing which capabilities to inject in a service, if we really want to optimize our services by providing only the relevant capabilities instead of all of them then we should use a dedicated factory.

dantleech commented on lib/LanguageServerCompletion/LanguageServerCompletionExtension.php:
```diff
@@ -40,33 +40,35 @@ public function configure(Resolver $schema)
                 $container->get(CompletionExtension::SERVICE_REGISTRY),
-                $container->get(SuggestionNameFormatter::class)
+                $container->get(SuggestionNameFormatter::class),
+                $capabilities['textDocument']['completion']['completionItem']['snippetSupport'] ?? false
```
yeah this is "fine" for now :) :+1:
